CommonItem/src/main/java/com/github/sef24sp4/common/item/loottable/LootTable.java

```diff
@@ -0,0 +1,89 @@
+package com.github.sef24sp4.common.item.loottable;
+
+import com.github.sef24sp4.common.item.CommonItem;
+import com.github.sef24sp4.common.item.ItemRarity;
+import com.github.sef24sp4.common.item.ItemSPI;
+
+import java.util.*;
+
+public class LootTable {
+	private final Map<ItemRarity, Double> itemChances;
+	private final Map<ItemRarity, Collection<ItemSPI>> providers;
+
+	private static Map<ItemRarity, Collection<ItemSPI>> generateItemProvidersMap(final Collection<ItemSPI> providers) {
+		final Map<ItemRarity, Collection<ItemSPI>> map = new HashMap<>();
+		for (ItemSPI provider : providers) {
+			map.computeIfAbsent(provider.getRarity(), k -> new ArrayList<>()).add(provider);
+		}
+		return Collections.unmodifiableMap(map);
+	}
+	/**
+	 * Constructs a new LootTable with the given item chances.
+	 * The item providers are loaded using the ServiceLoader mechanism.
+	 *
+	 * @param itemChances a map which maps {@link ItemRarity} to their chances. Chances are of type double.
+	 * @throws IllegalArgumentException if the sum of the chances is greater than 1, or any chance is negative or greater than 1
+	 * @see ItemRarity
+	 */
+	public LootTable(Map<ItemRarity, Double> itemChances) {
+		this(itemChances, ServiceLoader.load(ItemSPI.class).stream().map(ServiceLoader.Provider::get).toList());
+	}
+
+	/**
+	 * Constructs a new LootTable with the given item chances and item providers. Only used for testing.
+	 *
+	 * @param itemChances a map from {@link ItemRarity} to their chances of type double.
+	 * @param itemProviders a collection of item providers
+	 * @throws IllegalArgumentException if the sum of the chances is greater than 1, or any chance is negative or greater than 1
+	 * @see ItemRarity
+	 */
+	LootTable(Map<ItemRarity, Double> itemChances, final Collection<ItemSPI> itemProviders) {
+		this.providers = generateItemProvidersMap(itemProviders);
+		double sumChance = 0;
+		for (Map.Entry<ItemRarity, Double> entry : itemChances.entrySet()) {
+			double chance = entry.getValue();
+			if (chance < 0) throw new IllegalArgumentException("The chance of " + entry.getKey() + " is negative");
+			if (chance > 1) throw new IllegalArgumentException("The chance of " + entry.getKey() + " is greater than 1");
+			sumChance += chance;
+		}
+		if (sumChance > 1) throw new IllegalArgumentException("The sum of " + itemChances.keySet() + " is greater than 1");
```

Comment:
Please extract this to a private static helper method, isolate the validation logic from general constructor logic.